crates/oxc_linter_plugin/src/plugin.rs

```diff
@@ -0,0 +1,615 @@
+    #[error(transparent)]
+    Ignore(ignore::Error),
+    #[error(transparent)]
+    ReadFile(std::io::Error),
+    #[error("Failed to parse file at path: {0}")]
+    QueryParse(PathBuf, #[related] Vec<ParseError>),
+    #[error(
+        "Expected span_start and span_end to be List of Int or Int, instead got\nspan_start = {span_start}\nspan_end = {span_end}"
+    )]
+    WrongTypeForSpanStartSpanEnd {
+        span_start: String,
+        span_end: String,
+        #[source_code]
+        query_source: Arc<NamedSource>,
+        #[label = "This query failed."]
+        query_span: SourceSpan,
+    },
+}
+
+impl LinterPlugin {
+    pub fn new(schema: &'static Schema, queries_path: &PathBuf) -> oxc_diagnostics::Result<Self> {
+        let mut deserialized_queries = vec![];
+
+        for dir_entry_found_maybe in Walk::new(queries_path) {
+            let dir_entry_found = dir_entry_found_maybe.map_err(ErrorFromLinterPlugin::Ignore)?;
+
+            let pathbuf = dir_entry_found.path().to_path_buf();
+
+            if pathbuf.extension().map_or(false, |ext| ext.eq_ignore_ascii_case("yml")) {
+                let text = fs::read_to_string(&pathbuf).map_err(ErrorFromLinterPlugin::ReadFile)?;
+
+                let mut deserialized =
+                    serde_yaml::from_str::<InputQuery>(&text).map_err(|err| {
+                        ErrorFromLinterPlugin::QueryParse(pathbuf.clone(), vec![ParseError(err)])
+                    })?;
+                deserialized.path = pathbuf;
+                deserialized_queries.push(deserialized);
+            }
+        }
+
+        Ok(Self { rules: deserialized_queries, schema })
+    }
+
+    #[allow(clippy::too_many_lines)]
+    pub fn run_plugin_rules(
+        &self,
+        ctx: &mut LintContext,
+        plugin: &InputQuery,
+        adapter: &Adapter<'_>,
+        mv8: &mini_v8::MiniV8,
+    ) -> oxc_diagnostics::Result<()> {
+        let arc_adapter = Arc::new(adapter);
+        let query_source =
+            Arc::new(NamedSource::new(plugin.path.to_string_lossy(), plugin.query.clone()));
+        // NOTE: the 0 is technically wrong, but it's the right file which is enough
+        let query_span = SourceSpan::new(0.into(), plugin.query.len().into());
+
+        for data_item in
+            execute_query(self.schema, Arc::clone(&arc_adapter), &plugin.query, plugin.args.clone())
+                .map_err(|err| ErrorFromLinterPlugin::Trustfall {
+                    error_message: err.to_string(),
+                    query_source: Arc::clone(&query_source),
+                    query_span,
+                })?
+        {
+            let transformer: mini_v8::Function = mv8
+                .eval(plugin.post_transform.clone().unwrap_or_else(|| "(data) => data".to_string()))
+                .unwrap_or_else(|err| panic!("{err}"));
+
+            let data = mv8.create_object();
+
+            for (k, v) in data_item.clone() {
+                fn to_v8(mv8: &mini_v8::MiniV8, fv: FieldValue) -> Value {
```

Comment:
Having this inline function definition inside a huge function is not good practice. It causes poor readability and maintainability. Pull it out to the top level scope of the file and pass values into it explicitly.